### PR TITLE
Convert Filebeat iis.access to ECS

### PR DESCRIPTION
**Pull request by webmat**
Caveats

- [ ] user agent is encoded in the log. Decoding to get better UA parsing results.
  * It's not quoted, so spaces are replaced with `+` signs. Parens and slashes are not replaced, though.
  * Using `urldecode` on it makes it more palatable to `user_agent` parser.

Renames

- iis.access.server_ip => destination.ip
- iis.access.remote_ip => source.ip
- iis.access.method => http.request.method
- iis.access.url => url.path
- iis.access.query_string => url.query
- iis.access.port => destination.port
- iis.access.user_name => user.name
- iis.access.referrer => http.request.referrer
- iis.access.response_code => http.response.status_code
- iis.access.hostname => destination.domain
- iis.access.user_agent.original => user_agent.original
- iis.access.geoip => source.geo

TODO

- [x] Convert status code, port, byte counts, request times to int
- [x] Update ECS-migration.yml file
- [x] Changelog
- [x] Create field aliases
- [x] Output `user_agent` to ECS field names
  - [x] Next step: figure out what to do with fully broken down OS version numbers, when ECS supports one `version` string :-)
 - [x] List decoding of user agent in breaking changes

### CODE REVIEW

File: filebeat/module/iis/access/test/test.log-expected.json [hunk 1]
```diff
@@ -1,105 +1,105 @@
...
-        "iis.access.geoip.location.lat": 52.4908, 
-        "iis.access.geoip.location.lon": 13.3275, 
-        "iis.access.geoip.region_iso_code": "DE-BE", 
-        "iis.access.geoip.region_name": "Land Berlin", 
-        "iis.access.method": "GET", 
-        "iis.access.port": "80", 
-        "iis.access.query_string": "q=100", 
-        "iis.access.referrer": "-", 
-        "iis.access.remote_ip": "85.181.35.98", 
+        "http.request.method": "GET",
```
Comment (ruflin) [inline filebeat/module/iis/access/test/test.log-expected.json]: We should probably add a note to ECS that this should always be upper case.

Comment (webmat) [reply filebeat/module/iis/access/test/test.log-expected.json]: Good idea

**Comment by webmat** [comment]
@ruflin Ready for review.

File: filebeat/module/iis/access/test/test.log-expected.json (hunk 1 above)
Comment (webmat) [reply filebeat/module/iis/access/test/test.log-expected.json]: Added a note to the planning. Resolving this comment

File: filebeat/module/iis/access/ingest/default.json [hunk 2]
```diff
@@ -37,12 +37,12 @@
   }, {
     "rename": {
       "field": "iis.access.agent",
-      "target_field": "iis.access.user_agent.original"
+      "target_field": "user_agent.original"
```
[the remark this answers is not in the document]
Comment (webmat) [reply filebeat/module/iis/access/ingest/default.json]: This is a leftover from my attempts at outputting the UA processor's output directly to `user_agent`, which overwrites whatever's at `user_agent`, it doesn't merge results around `user_agent.original`, if already there.

I'll play with this some more, and see what the possibilities are, if we can get rid of that temporary field.

**Comment by webmat** [comment]
@ruflin I've worked a bit on the user_agent parsing. You were right about the mapping issue. So this is resolved. My work has revealed a few things:

- I noticed that in IIS logs, the UA is not quoted, and they therefore replace spaces with the `+`. I'm adding a `urldecode` step to get more precise UA parsing. (introduced in 6.1)
- I added a different UA in the test log, which does get it's OS major and minor version parsed successfully. This revealed a problem, outlined at next point.
- UA parser gives us fully broken down version fields (major, minor, patch), but no whole "version" field, whereas in ECS we only define "version".
  - This is showing up as a problem for user_agent.os.version. For now I haven't fixed this issue yet, I only created fields temporarily, so the tests don't break for this.

I think there's many ways we can go about this.

- ECS should likely add support for version breakdown fields
- Tackling the user_agent in this PR is informative. But we may want to close the other PRs without the UA fix, and do a subsequent PR only about getting UA right in all access logs.

**Comment by webmat** [comment]
This is ready for a final review.

Please make sure to check out the caveat above. I've made sure to list this "breaking" change improvement separately from the ECS transition.

File: filebeat/module/iis/access/ingest/default.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (webmat) [reply filebeat/module/iis/access/ingest/default.json]: Looping back on this: we need the intermediary field.

**Comment by ruflin** [comment]
For the version: I prefer the full version string. We had it broken down in an early version of ECS but I didn't see the benefits. Elasticsearch offers prefix queries which means queries for `major.minor` as an example are possible. The only thing not possible is for example `> major` but I would rather see ES have a type `version` which would support all these type of queries.

File: filebeat/module/iis/access/_meta/fields.yml [hunk 3]
```diff
@@ -82,87 +37,91 @@
...
-          description: >
-            The name of the operating system.
+          type: alias
+          path: user_agent.os.full_name
         - name: os_major
-          type: long
-          description: >
-            The major version of the operating system.
+          type: alias
+          path: user_agent.os.major
```
Comment (ruflin) [inline filebeat/module/iis/access/_meta/fields.yml]: We could probably use script processor to concatenate the version fields. Initially hoped the the join processor would do this but seems it has a different purpose.

Comment (webmat) [reply filebeat/module/iis/access/_meta/fields.yml]: Yeah, but all of this refining needs to go in a shared pipeline. We're not going to duplicate this work everywhere, on the eve of starting to share pipelines ;-)

**Comment by webmat** [comment]
Sorry, had missed this comment https://github.com/elastic/beats/pull/9084#issuecomment-441116257 before reading your review.

Reconstructing the version string is out of scope for this PR, IMO. The `user_agent` IN processor doesn't give it to us, at this time. I'm down with requesting that change to the IN team, or reconstructing it, but I don't think it should block this mapping to ECS.

I've added this to #9208 so we don't lose track of this.

**Pull request merged**